### ISSUE TITLE
Some inconsistencies in 2.4.4

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1479]
**Reported by beutlich on 13 May 2014 08:03 UTC**
I observed some inconsistencies in 2.4.4 on Strings.

In the table:
 * \n new-line: Better write: newline
 * \r return: Better write: carriage return
In the following note:
 * There is an opening parenthesis without matching closing parenthesis.
 * on Linux systems to have a “newline” character: Bettwer write: on Linux systems to have a line feed character
 * on Windows systems to have a “newline” and a “carriage return” character: Better write: on Windows systems to have a carriage return and a line feed character
 * What about Mac/OSX systems?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1479

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 13 May 2014 08:36 UTC**
Mac OSX is a Unix OS, and like Linux it uses LF.
Mac OS uses CR.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 19 Aug 2014 08:26 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Nov 2014 09:22 UTC**
Replying to [ticket:1479 beutlich]:
> I observed some inconsistencies in 2.4.4 on Strings.
> 
> In the table:
>  \* \n new-line: Better write: newline
>  \* \r return: Better write: carriage return
> In the following note:
>  \* There is an opening parenthesis without matching closing parenthesis.
>  \* on Linux systems to have a “newline” character: Bettwer write: on Linux systems to have a line feed character

Even if technically correct wouldn't that cause another inconsistency?

We first say that we have \n newline and \r carriage return (which are the names in C and we follow C-conventions for them), and then say that files on Linux systems contain a line feed character (which is the ASCII name for \n).

If we don't want to say that Linux files have newline and Windows files carriage return followed by newline, I would prefer that we used \n and \r\n respectively - since it is clear and avoids mixing C-names and ASCII-names. I assume there might be some deep difference between line feed and newline, but I don't see that it is important in that chapter.

(I believe the parenthesis should end at the end of the paragraph, or we should just remove the opening parenthesis.)

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 28 Nov 2014 09:33 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 28 Nov 2014 15:25 UTC**
Carriage return and line feed are characters, whereas newline is a concept. It just so happens that in Unix/Linux files, a newline is represented both in strings and in text files as a line feed ("\n"), whereas in DOS/Windows files only, a newline is represented as a carriage return followed by a line feed ("\r\n").

Most C runtime libraries for DOS/Windows further muddy the waters by saving a line feed as a carriage return / line feed pair when writing to a text file, and converting the same back into a single line feed on reading from the file.

So, technically "\n" is the input syntax for a line feed character, which is used to represent a newline in a string in memory, and in Unix/Linux systems, also in files. And "\r" is the input syntax for a carriage return character, which represents only itself.

In short, we can thank Unix for mixing up characters with concepts. When reading the text in 2.4.4, we have to think about which one we mean at any given point.

[_Aside: That Unix/Linux uses only line feed in files to represent a newline was probably done just to save space. The output hardware at the time required both a carriage return (to move the print head back to the left margin) and a line feed (to advance the paper by one line) in order to start a new line._
```
One two three\n
seven\n      four five six\r
eight nine\r\n
ten.
```
]

[_Aside2: Older versions of MacOS used only a "\r" to represent a newline in a file. Probably for no other reason than "it's not what DOS or Unix does"._]

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Nov 2014 15:55 UTC**
Replying to [comment:5 stefanv]:
> Carriage return and line feed are characters, whereas newline is a concept. It just so happens that in Unix/Linux files, a newline is represented both in strings and in text files as a line feed ("\n"), whereas in DOS/Windows files only, a newline is represented as a carriage return followed by a line feed ("\r\n").

Thanks for the explanation, I also tried to read a bit more.
However, I think this makes it clearer that for the explanation it seems best to just write:

Note, if the contents of a file is read into a Modelica string, it is assumed that the reading function is responsible to handle the different line ending symbols on file, e.g. on Linux systems to have a “\n” character at the end of a line and on Windows systems to have the “\r\n” characters. As usual in programming languages, the content of a file in a Modelica string only contains the “\n” character.

> Most C runtime libraries for DOS/Windows further muddy the waters by saving a line feed as a carriage return / line feed pair when writing to a text file, and converting the same back into a single line feed on reading from the file.

Well, in their defense they follow the C-standard!
 
> So, technically "\n" is the input syntax for a line feed character, which is used to represent a newline in a string in memory, and in Unix/Linux systems, also in files. And "\r" is the input syntax for a carriage return character, which represents only itself.
> 
> In short, we can thank Unix for mixing up characters with concepts. When reading the text in 2.4.4, we have to think about which one we mean at any given point.
> 
> [_Aside: That Unix/Linux uses only line feed in files to represent a newline was probably done just to save space. The output hardware at the time required both a carriage return (to move the print head back to the left margin) and a line feed (to advance the paper by one line) in order to start a new line._

Yes, that seems consistent with the rest of the Unix philosophy.

Aside3: As far as I understand some hardware was also too slow to move the head back quick enough (for just "\r") and thus used the "\n" as a delay.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 28 Nov 2014 16:02 UTC**
Replying to [comment:6 hansolsson]:
> As usual in programming languages, the content of a file in a Modelica string only contains the “\n” character.

Well, we might want to allow _some_ other characters than only '\n'. Should probably formulate that slightly differently.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 28 Nov 2014 16:21 UTC**
Hans, I don't think the C standard says anything about mapping of "\n" to "\r\n" when writing to files. I don't have a copy of the standard, but my 2nd Edition K&R talks only of the "\r" and "\n" characters separately. In the table in the appendix (p.193) are these two entries:

```
newline          NL (LF)  \n
...
carriage return  CR       \r
```

I think the Modelica specification should state something along the lines of:

    The "\n" character is used to conceptually indicate the end of a line within a Modelica string. Any Modelica program that needs to recognize line endings can check for a single "\n" character to do so on any platform. It is the responsibility of a Modelica implementation to make any necessary transformations to other representations when writing to or reading from a text file.

    [_For example, a "\n" is written and read as-is in a Unix or Linux implementation, but written as "\r\n" pair, and converted back to "\n" when read, in a Windows implementation._]

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Dec 2014 11:21 UTC**
Language group: Stefan's variant is ok (except some discussion about changing "program" to "code"/"algorithm"/"function").

Accept "program".

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Dec 2014 10:05 UTC**
Language group:
Clarify that "\n" and literal newline in a string are treated the same way; by adding "same as literal newline" after "\n" in the table.

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 10 Dec 2014 15:40 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 10 Dec 2014 15:46 UTC**
Decision was made at design meeting. Now need to update spec.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 16 Jul 2015 15:48 UTC**
Resolved in r8348
